diff -Naur gettext-0.18.3/gettext-tools/gnulib-lib/getdtablesize.c ../Source/gettext-0.18.3/gettext-tools/gnulib-lib/getdtablesize.c
--- gettext-0.18.3/gettext-tools/gnulib-lib/getdtablesize.c	2013-07-07 02:16:17.000000000 -0700
+++ ../Source/gettext-0.18.3/gettext-tools/gnulib-lib/getdtablesize.c	2015-12-18 17:34:06.000000000 -0800
@@ -82,5 +82,19 @@
     }
   return dtablesize;
 }
+#elif defined(__ANDROID__)
 
+#include <sys/resource.h>
+#include <sys/sysconf.h>
+
+int
+getdtablesize (void)
+{
+  struct rlimit r;
+
+  if (getrlimit(RLIMIT_NOFILE, &r) < 0) {
+    return sysconf(_SC_OPEN_MAX);
+  }
+  return r.rlim_cur;
+}
 #endif
diff -Naur gettext-0.18.3/gettext-tools/libgrep/nl_langinfo.c ../Source/gettext-0.18.3/gettext-tools/libgrep/nl_langinfo.c
--- gettext-0.18.3/gettext-tools/libgrep/nl_langinfo.c	2013-07-07 02:17:02.000000000 -0700
+++ ../Source/gettext-0.18.3/gettext-tools/libgrep/nl_langinfo.c	2015-12-18 11:40:22.000000000 -0800
@@ -16,6 +16,7 @@
    along with this program.  If not, see <http://www.gnu.org/licenses/>.  */
 
 #include <config.h>
+#include <locale.h>
 
 /* Specification.  */
 #include <langinfo.h>
@@ -152,10 +153,12 @@
       return "ISO-8859-1";
 # endif
     /* nl_langinfo items of the LC_NUMERIC category */
+/*
     case RADIXCHAR:
       return localeconv () ->decimal_point;
     case THOUSEP:
       return localeconv () ->thousands_sep;
+*/
     /* nl_langinfo items of the LC_TIME category.
        TODO: Really use the locale.  */
     case D_T_FMT:
@@ -255,8 +258,10 @@
       return "\0\0\0\0\0\0\0\0\0\0";
     /* nl_langinfo items of the LC_MONETARY category
        TODO: Really use the locale. */
+/*
     case CRNCYSTR:
       return "-";
+*/
     /* nl_langinfo items of the LC_MESSAGES category
        TODO: Really use the locale. */
     case YESEXPR:
diff -Naur gettext-0.18.3/gettext-tools/src/msginit.c ../Source/gettext-0.18.3/gettext-tools/src/msginit.c
--- gettext-0.18.3/gettext-tools/src/msginit.c	2012-12-03 22:28:58.000000000 -0800
+++ ../Source/gettext-0.18.3/gettext-tools/src/msginit.c	2015-12-16 11:00:47.000000000 -0800
@@ -1081,7 +1081,11 @@
       char *result;
 
       /* Return the pw_gecos field, up to the first comma (if any).  */
+#ifndef __ANDROID__
       fullname = pwd->pw_gecos;
+#else
+      fullname = "android";
+#endif
       fullname_end = strchr (fullname, ',');
       if (fullname_end == NULL)
         fullname_end = fullname + strlen (fullname);
